src/core_plugins/kibana/public/dashboard/dashboard.js

```diff
@@ -180,104 +142,81 @@ app.directive('dashboardApp', function (Notifier, courier, AppState, timefilter,
        * @returns {Object}
        */
       $scope.createChildUiState = function createChildUiState(path, uiState) {
-        return $scope.uiState.createChild(path, uiState, true);
-      };
-
-      $scope.saveState = function saveState() {
-        $state.save();
+        return dashboardState.uiState.createChild(path, uiState, true);
       };
 
       $scope.onPanelRemoved = (panelIndex) => {
-        _.remove($scope.state.panels, function (panel) {
+        _.remove(dashboardState.getPanels(), function (panel) {
           if (panel.panelIndex === panelIndex) {
-            $scope.uiState.removeChild(getPersistedStateId(panel));
+            dashboardState.uiState.removeChild(getPersistedStateId(panel));
             return true;
           } else {
             return false;
           }
         });
-        $state.save();
+        dashboardState.saveState();
       };
 
-      $scope.brushEvent = brushEvent;
-      $scope.filterBarClickHandler = filterBarClickHandler;
-      $scope.expandedPanel = null;
-      $scope.hasExpandedPanel = () => $scope.expandedPanel !== null;
-      $scope.toggleExpandPanel = (panelIndex) => {
-        if ($scope.expandedPanel && $scope.expandedPanel.panelIndex === panelIndex) {
-          $scope.expandedPanel = null;
-        } else {
-          $scope.expandedPanel =
-            $scope.state.panels.find((panel) => panel.panelIndex === panelIndex);
-        }
+      $scope.save = function () {
+        // Make sure to save the latest query, even if 'enter' hasn't been hit.
+        dashboardState.updateFilters(queryFilter);
+        return dashboardState.saveDashboard(angular.toJson).then(function (id) {
+          $scope.kbnTopNav.close('save');
+          if (id) {
+            notify.info(`Saved Dashboard as "${dash.title}"`);
+            if (dash.id !== $routeParams.id) {
+              kbnUrl.change(
+                `${DashboardConstants.EXISTING_DASHBOARD_URL}`,
+                { id: dash.id });
+            } else {
+              docTitle.change(dash.lastSavedTitle);
+            }
+          }
+        }).catch(notify.fatal);
       };
 
+      $scope.$watchCollection('options', function (newVal, oldVal) {
+        if (!angular.equals(newVal, oldVal)) {
+          dashboardState.saveState();
```

Comment:
The fact that we watch some value `options` on the scope and in reaction call some function on `dashboardState` suggests that there is a "hidden" complicating factor that can be pain in later refactorings. We could try to make our future selves happier by watching `() => dashboard.getOptions()` instead, which would obviously state: "If the options in the dashboard state have changed, save the state".